### PR TITLE
Make it possible to disable Flatten feature for a given report

### LINKED ISSUE #11323: Scheduled PDF Reports: Report Referrer Type shows only keywords but not detailed titles
The referrer type report shows only keywords like search engines, websites, ... for the entries. It should show detailed titles.
grafik

### PR DESCRIPTION
**Pull request by sgiehl**
Currently datatables are automatically flattened if the `flat` parameter is given in the request. This cannot be prevented.
It is now possible to set for a report if it supports flattening or not. This will also automatically be set to the view, so it's not required to set if for report and view as well. 

This PR also disables the flattening for the referrer type report, as it doesn't work for that report.

fixes #11323

### CODE REVIEW

File: core/DataTable/Filter/Flatten.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        $flattener = new Flattener();
+        if ($this->includeAggregateRows) {
+            $flattener->includeAggregateRows();
+        }
+
+        $newTable = $flattener->flatten($table, $this->recursiveLabelSeparator);
+
+        // replace rows and metadata with those of the flattened datatable
+        $table->setRows($newTable->getRows());
+        $table->setAllTableMetadata($newTable->getAllTableMetadata());
```
Comment (sgiehl) [inline core/DataTable/Filter/Flatten.php]: Replacing the whole datatable in an filter is not possible. Guess replacing all rows and metadata should be almost the same.

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,7 @@ The Product Changelog at **[piwik.org/changelog](http://piwik.org/changelog)** l
 
 ### Breaking Changes
 * New config setting `enable_plugin_upload` lets you enable uploading and installing a Piwik plugin ZIP file by a Super User. This used to be enabled by default, but it is now disabled by default now for security reasons.
+* New property `Report::$supportsFlattening` lets you define if a report supports flattening (d)efaults to `true`). If set to `false` it will also set `ViewDataTable\Config::$show_flatten_table` to `false`
```
Comment (mattab) [inline CHANGELOG.md]: Indicate what "property" refers to. something like `New Report class property` or so

File: core/Plugin/Report.php [hunk 3]
```diff
@@ -120,6 +120,14 @@ class Report
     protected $hasGoalMetrics = false;
 
     /**
+     * Set this property to false in case your report can't/shouldn't be flattened.
+     * In this case, flattener won't be applied even if parameter is provided in a request
+     * @var bool
+     * @api
+     */
+    protected $supportsFlattening = true;
```
Comment (mattab) [inline core/Plugin/Report.php]: rename to `supportsFlatten`

File: core/Plugin/Report.php [hunk 4]
```diff
@@ -494,6 +502,15 @@ public function hasGoalMetrics()
     }
 
     /**
+     * @return bool
+     * @ignore
+     */
+    public function supportsFlattening()
```
Comment (mattab) [inline core/Plugin/Report.php]: rename to `supportsFlatten`

**Review by mattab — COMMENTED** [review]
Looks good. Left a couple naming comments

**Comment by mattab** [comment]
:+1:

**Pull request merged**